### PR TITLE
Improve type definitions for the viewer

### PR DESCRIPTION
**Pull request by ex37**
This commit improves compatibility of the viewer code with TypeScript by including missing type imports/definitions and correcting existing ones

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Given that the default viewer is *not* intended to be directly used by third-parties, and thus isn't included in `pdfjs-dist`, I don't know if this is really desirable. Please see http://mozilla.github.io/pdf.js/getting_started/#introduction (emphasis mine):
> The viewer is built on the display layer and is the UI for PDF viewer in Firefox and the other browser extensions within the project. It can be a good starting point for building your own viewer. ***However, we do ask if you plan to embed the viewer in your own site, that it not just be an unmodified version. Please re-skin it or build upon it.***

**Comment by ex37** [comment]
@Snuffleupagus this could help people who want to extend the viewer. Plus there are also lot's of nice utils within the web folder. Also added jsdoc types could be helpful for project's contributors.

But if you think this is not within the scope of the project, we can remove the entrypoint, and just add the type fixes.

**Comment by Snuffleupagus** [comment]
> But if you think this is not within the scope of the project, we can remove the entrypoint, and just add the type fixes.

That'd be a requirement as far as I'm concerned. Please note that the existing type definitions are unused within the PDF.js project, but has nonetheless added a bunch of work for the core contributors; hence we don't want to further increase that.

Also, please make sure to write a good *commit message*.

File: web/annotation_editor_params.js
Comment (Snuffleupagus) [inline web/annotation_editor_params.js:18]: Here, and everywhere else throughout the patch, imported typedefs should be placed *before* the regular `import`s.

File: web/base_tree_viewer.js
Comment (Snuffleupagus) [inline web/base_tree_viewer.js:46]: Here, and everywhere else throughout the patch, why do we need to make this change?

File: web/base_tree_viewer.js
Comment (Snuffleupagus) [inline web/base_tree_viewer.js:76]:
```suggestion
   * @param {HTMLDivElement} div
   * @param {boolean} [hidden]
   * @protected
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/base_tree_viewer.js
Comment (ex37) [reply web/base_tree_viewer.js:46]: cause typescript wise you can't redefine a private property in a child class (was leading to a broken .d.ts)

File: web/base_tree_viewer.js
Comment (ex37) [reply web/base_tree_viewer.js:76]: this method's signature is different for child classes (they are passing an object as a second parameter). In typescript this is not allowed, that's why I set hidden to any as a workaround.

Comment (Snuffleupagus) [reply web/base_tree_viewer.js:76]: No, they are not:
 - https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_outline_viewer.js#L185
 - https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_layer_viewer.js#L103

Hence the changes I mentioned are correct here for *the base* method.

Comment (ex37) [reply web/base_tree_viewer.js:76]: the call to parent class is correct, but I am talking about the child classes' redefined method signature
https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_outline_viewer.js#L167

Comment (ex37) [reply web/base_tree_viewer.js:76]: here is a simple demo, just for case:

https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKYCcCmA7ALtA3gKGtAtkmgBYD2AJgBSIDmAXNCgK74BGSCAlFtrtAL7ZB2UJBgBhYgEsQ5aEgAeaVORjxk6HnkIkK1BPWitSpEEjApuOPHj6D+QA

File: tsconfig.json
Comment (ex37) [inline tsconfig.json:19]: @Snuffleupagus should I get rid of path aliases as well? Those have no use besides generating types, but was hard to figure out the list so could be helpful for someone else like myself.

Comment (Snuffleupagus) [reply tsconfig.json:19]: Yes, given that we don't want to add unused code.

(Those shouldn't need to be figured out, since they can basically be copied from `gulpfile.mjs`.)

**Review by ex37 — COMMENTED** [review]
Addressed requested changes, changed commit message / pr name.
Also added `eslint-disable-next-line` where needed.

File: web/annotation_editor_params.js
Comment (Snuffleupagus) [inline web/annotation_editor_params.js:27]:
```suggestion
 * @property {HTMLButtonElement} editorStampAddImage
```

File: web/annotation_editor_params.js
Comment (Snuffleupagus) [inline web/annotation_editor_params.js:29]:
```suggestion
 * @property {HTMLButtonElement} editorHighlightShowAll
```

File: web/external_services.js
Comment (Snuffleupagus) [inline web/external_services.js:16]:
```suggestion
/** @typedef {import("./interfaces.js").IL10n} IL10n */
```

File: web/external_services.js
Comment (Snuffleupagus) [inline web/external_services.js:34]:
```suggestion
   * @returns {Promise<IL10n>}
```

File: web/pdf_attachment_viewer.js
Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js:34]:
```suggestion
 * @property {boolean} [keepRenderedCapability]
```

File: web/pdf_layer_viewer.js
Comment (Snuffleupagus) [inline web/pdf_layer_viewer.js:19]: That file doesn't even exist, hence this cannot possibly work.
```suggestion
/** @typedef {import("../src/display/api").PDFDocumentProxy} PDFDocumentProxy
```

File: web/pdf_outline_viewer.js
Comment (Snuffleupagus) [inline web/pdf_outline_viewer.js:20]:
```suggestion
/** @typedef {import("../src/display/api").PDFDocumentProxy} PDFDocumentProxy
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

**Comment by Snuffleupagus** [comment]
https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

**Comment by ex37** [comment]
squashed

File: web/grab_to_pan.js
Comment (Snuffleupagus) [inline web/grab_to_pan.js:23]: That's not a pattern we use anywhere in the code-base really.
It seems more correct to, just above, add
```javascript
/**
 * @typedef {Object} GrabToPanOptions
 * @property {HTMLElement} element
 */
```
and then write this as follows
```suggestion
   * @param {GrabToPanOptions} options
```

**Comment by timvandermeij** [comment]
Before we can merge this please also include a better commit message. The guidelines at https://cbea.ms/git-commit/ help with this. The commit message should primarily show which change is being made and why in e.g. `git log` output; implementation details like which files or classes are changed can usually be left out because that's already clear from looking at the commit diff. A suggestion could be:

```
Improve type definitions for the viewer

This commit improves compatibility of the viewer code with TypeScript by
including missing type imports/definitions and correcting existing ones.
```

**Comment by ex37** [comment]
Got it. Set to your suggested description, as it fully describes what this pr is about.

File: web/pdf_outline_viewer.js
Comment (timvandermeij) [inline web/pdf_outline_viewer.js:19]: Missing `.js` extension here as well.

**Review by timvandermeij — APPROVED** [review]
Looks good to me with the comments below addressed, but given the original review I'll leave the final approval to @Snuffleupagus. Thanks!

File: web/pdf_document_properties.js
Comment (ex37) [reply web/pdf_document_properties.js:19]: it works for typescript, yes, but for consistency makes sense to add extensions everywhere

File: web/base_tree_viewer.js
Comment (ex37) [reply web/base_tree_viewer.js:75]: it's required, as discussed here https://github.com/mozilla/pdf.js/pull/17879#discussion_r1549981654

**Comment by ex37** [comment]
removed the default value for `hidden`, and also added missing extensions

File: web/base_tree_viewer.js
Comment (timvandermeij) [reply web/base_tree_viewer.js]: Huh, that's interesting and something we might want to change later, but indeed not in this PR. Let's keep the object part of this comment as-is then.

**Review by timvandermeij — APPROVED** [review]

**Comment by ex37** [comment]
@timvandermeij in my opinion a proper fix for `_addToggleButton` method would probably be separating it into two methods, which will be using same signature across inheriting classes. Other than type definitions, in my understanding, it can also help with runtime optimizations.  But yes, that's outside of this pr scope.

**Review by Snuffleupagus — APPROVED** [review]
r=me, thanks.

**Pull request merged**
